### PR TITLE
[Test refactor 3/5] Notification service improvement

### PR DESCRIPTION
**Pull request by LysandreJik**
This improves the notification service in Slack to provide a more comprehensive view over the tests.

Previously:

image

Now:

image

To do before merge:
- Update the slack channel reference so that it posts to the right channel
- Ensure compatibility with the `push` job that runs on each commit to `master`

### CODE REVIEW

**Comment by LysandreJik** [comment]
Formatting it as you offer isn't complicated, it was using pandas that I wasn't too keen on. I personally prefer the format above but it's probably because I have seen it too much by now :) Let's hear what others think.

And indeed, will update the categories at the top.

**Review by patrickvonplaten — APPROVED** [review]

File: utils/notification_service.py [hunk 1]
```diff
@@ -38,183 +64,607 @@ def handle_test_results(test_results):
...
+
+        failures = {k: per_model_sum(v) for k, v in self.model_results.items() if sum(per_model_sum(v).values())}
+
+        model_reports = []
+        other_module_reports = []
+        for key, value in failures.items():
+            device_report = self.get_device_report(value, rjust=3)
+
+            if sum(value.values()):
+                report = f"{key}"
```
Comment (sgugger) [inline utils/notification_service.py:252]:
```suggestion
                report = key
```

File: utils/notification_service.py [hunk 2]
```diff
@@ -38,183 +64,607 @@ def handle_test_results(test_results):
...
+                else:
+                    model_reports.append(report)
+
+        model_failures_report = "\n".join(sorted(model_reports, key=lambda s: s.split("] ")[-1]))
+        module_failures_report = "\n".join(sorted(other_module_reports, key=lambda s: s.split("] ")[-1]))
+
+        return {
+            "type": "section",
+            "text": {
+                "type": "mrkdwn",
```
Comment (sgugger) [inline utils/notification_service.py:267]:
```suggestion
                "type": "markdown",
```
or is the typo intentional?

File: utils/notification_service.py [hunk 3]
```diff
@@ -38,183 +64,607 @@ def handle_test_results(test_results):
...
+                    report = f"[{device_report}] {report}"
+
+                individual_reports.append(report)
+
+        failures_report = "\n".join(sorted(individual_reports))
+
+        return {
+            "type": "section",
+            "text": {
+                "type": "mrkdwn",
```
Comment (sgugger) [inline utils/notification_service.py:296]: Same here. Seems intentional.

File: utils/notification_service.py [hunk 4]
```diff
@@ -38,183 +64,607 @@ def handle_test_results(test_results):
...
+
+    def post(self):
+        print("Sending the following payload")
+        print(json.dumps({"blocks": json.loads(self.payload)}))
+        self.thread_ts = client.chat_postMessage(
+            channel=os.environ["CI_SLACK_CHANNEL_DUMMY_TESTS"],
+            blocks=self.payload,
+            text=f"{self.n_failures} failures out of {self.n_tests} tests,"
+            if self.n_failures
+            else "All tests passed.",
```
Comment (sgugger) [inline utils/notification_service.py:353]: This deserves its own line before to be more readable.

File: utils/notification_service.py [hunk 5]
```diff
@@ -38,183 +64,607 @@ def handle_test_results(test_results):
...
+            failures = "\n".join(failures.split("\n")[:20]) + "\n\n[Truncated]"
+
+        title = job_name
+        if device is not None:
+            title += f" ({device}-gpu)"
+
+        content = {
+            "type": "section",
+            "text": {"type": "mrkdwn", "text": text},
+        }
```
Comment (sgugger) [inline utils/notification_service.py:367]: Can probably fit in one line :-)

**Review by sgugger — APPROVED** [review]
Thanks a lot for revamping this!

File: utils/notification_service.py (hunk 2 above)
Comment (LysandreJik) [reply utils/notification_service.py:267]: Yes, that's what the Slack sdk expects

File: utils/notification_service.py (hunk 3 above)
Comment (LysandreJik) [reply utils/notification_service.py:296]: Yes, that's what the Slack sdk expects

File: utils/notification_service.py [hunk 6]
```diff
@@ -12,13 +12,39 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import ast
+import collections
+import functools
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply utils/notification_service.py:17]: I hope the actual usage of this isn't black magic :innocent:

**Comment by LysandreJik** [comment]
I have updated the notification service to output the following format:

image

**Review by sgugger — APPROVED** [review]

**Pull request merged**